src/cli/abi.ts

```diff
@@ -0,0 +1,198 @@
+import ts, { Decorator, NodeArray } from "typescript";
+import * as abi from "near-abi";
+import * as TJS from "typescript-json-schema";
+import { JSONSchema7 } from "json-schema";
+import * as fs from "fs";
+import { LIB_VERSION } from "../version.js";
+
+function parseMetadata(packageJsonPath: string): abi.AbiMetadata {
+    const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
+
+    let authors: string[] = [];
+    if (packageJson["author"]) authors.push(packageJson["author"]);
+    authors = authors.concat(packageJson["contributors"] || []);
+
+    return {
+        name: packageJson["name"],
+        version: packageJson["version"],
+        authors,
+        build: {
+            compiler: "tsc " + ts.version,
+            builder: "near-sdk-js " + LIB_VERSION
+        }
+    };
+}
+
+function getProgramFromFiles(
+    files: string[],
+    jsonCompilerOptions = {},
+    basePath = "./"
+): ts.Program {
+    // use built-in default options
+    const compilerOptions = ts.convertCompilerOptionsFromJson(jsonCompilerOptions, basePath).options;
+    const options: ts.CompilerOptions = {
+        noEmit: true,
+        emitDecoratorMetadata: true,
+        experimentalDecorators: true,
+        target: ts.ScriptTarget.ES5,
+        module: ts.ModuleKind.CommonJS,
+        allowUnusedLabels: true,
+    };
+    for (const k in compilerOptions) {
+        // eslint-disable-next-line no-prototype-builtins
+        if (compilerOptions.hasOwnProperty(k)) {
+            options[k] = compilerOptions[k];
+        }
+    }
+    return ts.createProgram(files, options);
+}
+
+export function runAbiCompilerPlugin(
+    tsFile: string,
+    packageJsonPath: string,
+) {
+    const program = getProgramFromFiles([tsFile]);
+    const typeChecker = program.getTypeChecker();
+
+    const diagnostics = ts.getPreEmitDiagnostics(program);
+    if (diagnostics.length > 0) {
+        diagnostics.forEach((diagnostic) => {
+            const message = ts.flattenDiagnosticMessageText(diagnostic.messageText, "\n");
+            if (diagnostic.file) {
+                const { line, character } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start!);
+                console.error(`${diagnostic.file.fileName} (${line + 1},${character + 1}): ${message}`);
+            } else {
+                console.error(message);
+            }
+        });
+        return undefined;
+    }
```

Comment:
We can get rid of the preceding CLI step that does typechecking since this piece does effectively the same without having to instantiate `tsc` twice. But up to you guys.

Reply:
It is still unclear to me if IBI is a separate package. If it is not - then we can do a TS check once.